Implement TLS-Exporter

mfil opened the pull request:
## Description

This pull request implements the TLS-Exporter feature as defined in RFC 8446, Section 7.5 and RFC 5705.

TLS-Exporter allows the client and server to extract additional shared symmetric keys from the SSL context by inputting a label and a desired length for the key.

Currently, it is possible for library users to implement TLS-Exporter in TLS 1.2 by using `mbedtls_ssl_set_export_keys_cb()` to obtain the master secret and then calculate `mbedtls_ssl_tls_prf()`. It is not currently possible to do this for TLS 1.3. This pull request adds the function `mbedtls_ssl_export_keying_material()` to implement TLS-Exporter in the library for both TLS 1.2 and 1.3.

~~I have marked this pull request as "Draft" because I have not yet added any automated tests, and I would like some help in figuring out what would be the best way to add them. I *have* added options to `ssl_client2` and `ssl_server2` to print out the derived symmetric keys on the command line. I have checked that when connecting `openssl s_client` (with the `-keymatexport` option) to `ssl_server2`, they both produce the same key.~~

I have added a test for the TLS 1.3 Exporter. I could not find test vectors online, so I have taken the "exp master" key from RFC 8448 and used an online HMAC-SHA256 calculator to calculate the expected result. Additionally, I have added options to `ssl_client2` and `ssl_server2` to print out the derived symmetric keys on the command line. I have checked that when connecting `openssl s_client` (with the `-keymatexport` option) to `ssl_server2`, they both export the same key.

## PR checklist

Please remove the segment/s on either side of the | symbol as appropriate, and add any relevant link/s to the end of the line.
If the provided content is part of the present PR remove the # symbol.

- [x] **changelog** provided
- [x] **development PR** provided
- [x] **framework PR** not required
- [x] **3.6 PR** provided #9469
- **2.28 PR** not required because: TLS 1.3 is only experimental in this version, and the user can implement TLS-Exporter in TLS 1.2
- [x] **tests**  provided

mfil commented:
Sorry for the force-push. I forgot to sign off my latest commit, and there's no way to fix that with more commits.

I can close and reopen this merge request, but first I'd like to ask for some help with implementing unit tests.

davidhorstmann-arm commented:
@mfil Thanks for this very succinct and well-documented PR. @waleed-elmelegy-arm and I plan to do reviews over the next couple of months so we can try to get this merged.

In the meantime, there are some merge conflicts (I think from the 3.6 release). Would you be able to resolve these? We aren't planning much work in this area of the code for a while, so it will be unlikely to bitrot again.

mfil commented:
> @mfil Thanks for this very succinct and well-documented PR. @waleed-elmelegy-arm and I plan to do reviews over the next couple of months so we can try to get this merged.
> 
> In the meantime, there are some merge conflicts (I think from the 3.6 release). Would you be able to resolve these? We aren't planning much work in this area of the code for a while, so it will be unlikely to bitrot again.

I've rebased onto develop, fixed some mistakes in the comments, and force-pushed. (Let me know if you prefer opening a new pull request.)

I'll wait for reviews before changing the 3.6 backport pull request, ok?

mfil commented:
> > I'll wait for reviews before changing the 3.6 backport pull request, ok?
> 
> That sounds fine to me, thanks! I'll start the CI again, I seem to remember it failed the last time with a build error in some configurations. If you need a hand reproducing it let me know.

Most of them make sense to me.

In all_u16-check_names, it complains that `mbedtls_ssl_export_keying_material` isn't declared in a header, but it's right there in include/mbedtls/ssl.h, so I don't understand that.

gilles-peskine-arm commented on include/mbedtls/ssl.h:
```diff
@@ -5591,6 +5591,30 @@ int  mbedtls_ssl_tls_prf(const mbedtls_tls_prf_types prf,
+  *
+  * \return            0 on success. An SSL specific error on failure.
+  */
+ int mbedtls_ssl_export_keying_material(mbedtls_ssl_context *ssl,
```
> In all_u16-check_names, it complains that mbedtls_ssl_export_keying_material isn't declared in a header, but it's right there in include/mbedtls/ssl.h, so I don't understand that.

`check_names` is picky about the code layout. Here it's skipping the line because it's indented. Please remove the space and `check_names` should be happy.

mfil replied:
Huh, I indented the whole thing by one space. No idea how that happened.

mfil commented:
@davidhorstmann-arm I *think* I fixed everything. Can you re-run the CI?

gilles-peskine-arm commented:
I've started a CI run.

mfil commented:
> I've started a CI run.

Thank you!

I'm stuck on getting the last tests to succeed. The TLS 1.2 Exporter needs client_random and server_random. If I'm seeing this correctly, if `MBEDTLS_SSL_CONTEXT_SERIALIZATION` is not defined, then these go away when the handshake is done.

Would it be a problem to always have a `randbytes` field in `struct mbedtls_ssl_transform`?

tom-cosgrove-arm commented:
I've started the CI on the updated PR

mfil commented:
Ok, the code style check is happy now, that still leaves my problem I mentioned above.

mfil commented:
I've fixed the merge conflict and added a commit that should fix the remaining CI failures. Please re-run the CI @davidhorstmann-arm @gilles-peskine-arm

mfil commented:
Hooray!

davidhorstmann-arm commented on library/ssl_misc.h:
```diff
@@ -1161,10 +1161,10 @@ struct mbedtls_ssl_transform {
 #endif /* MBEDTLS_SSL_DTLS_CONNECTION_ID */
 
-#if defined(MBEDTLS_SSL_CONTEXT_SERIALIZATION)
+#if defined(MBEDTLS_SSL_CONTEXT_SERIALIZATION) || defined(MBEDTLS_SSL_PROTO_TLS1_2)
```
64 extra bytes is quite a lot to add into an LTS (assuming you'd like to get this into 3.6).

Would you be happy to instead only enable the TLS-Exporter when `MBEDTLS_SSL_CONTEXT_SERIALIZATION` is enabled? It is enabled in the default configuration so that shouldn't be too disruptive.

mfil replied:
It wouldn't bother me for my use case, but it would be very unintuitive.

Also: In TLS 1.3 we don't need `randbytes` for the exporter, so I wouldn't like to gate it behind `MBEDTLS_SSL_CONTEXT_SERIALIZATION` for TLS 1.3. But then what to do if we build with both TLS 1.2 and 1.3 enabled and serialization disabled?

That said, if we can't find a better solution then I'm ok with it.

mfil replied:
I made it so that `mbedtls_ssl_export_keying_material` exists if either `MBEDTLS_SSL_CONTEXT_SERIALIZATION` is defined or `MBEDTLS_SSL_PROTO_TLS1_2` is *not* defined. But maybe that's too convoluted?

What do you think about making a `MBEDTLS_SSL_ENABLE_EXPORT_KEYING_MATERIAL` define in mbedls_config.h instead?

davidhorstmann-arm replied:
I'm leaning towards the same conclusion that maybe a new config option is best. We try to avoid adding them if possible as it makes testing more difficult, but this should be fairly self-contained.

I'll have a think about it and get back to you tomorrow.

davidhorstmann-arm replied:
Having not come up with anything better, I agree. Could you add a config option in `mbedtls_config.h`?

Ideally, it would be off by default, but that won't work if you're consuming Mbed TLS binaries built in the default config, for example if you consume a distro package. If you need it to be on in the default config that is probably okay.

In any case it will be turned on in the `full` config.

I think `MBEDTLS_SSL_KEYING_MATERIAL_EXPORT` would be a good name (the 'enable' seems superfluous since that's true of nearly all config options and the order swap seems clearer to me).

mfil replied:
> Ideally, it would be off by default, but that won't work if you're consuming Mbed TLS binaries built in the default config, for example if you consume a distro package. If you need it to be on in the default config that is probably okay.

Off by default is fine by me. We're compiling and statically linking Mbed TLS. I don't think anyone besides me has asked for this feature on the mailing list, and if someone does, you could always change the default.

mfil replied:
I've added the option. Let me know what you think!

mfil replied:
I enabled it by default.

I could see a case for leaving it off by default in 3.6, someone might compile a custom version with serialization off, and then a patch release suddenly makes the SSL context larger, and they need to figure out where that comes from.

davidhorstmann-arm replied:
Usually, if a user is are compiling a custom configuration, they have their own separate config file which they leave unchanged when they upgrade. This means that if we add a new option it is automatically disabled in their config file (because their file doesn't set the new option).

The only situation where this could be a problem is if the user's upgrade process looks like:
1. Get the new version of the source tree
2. Disable a few listed things in `mbedtls_config.h`
3. Build

But I have not come across anything like this in the wild, and I don't think we make any guarantees about it working.

mfil replied:
That's actually kind of how we do it for [OpenVPN-NL](https://openvpn.fox-it.com). We've got a small number of patches we apply on top of mbed TLS, mostly changing the config file. Of course, since we do make some changes outside the config file, we can't assume any guarantees from you anyway!

Anyway, it doesn't bother me, so feel free to hit the resolve button!

davidhorstmann-arm replied:
Well in that case I stand corrected, I have seen it in the wild and will bear this in mind for future!

But as you say, I think this is the best thing to do in this case, I'll mark this as resolved.

davidhorstmann-arm reviewed (changes requested):
I've reviewed to a first approximation, thanks for your patience!

There are 2 other points I would like to address:

1. Re testing, it would be good to have more negative testing, particularly of error cases like buffer arguments too small or attempting to export too much key material from the PRF. The easiest way to do this would be to use the SSL helpers `mbedtls_test_ssl_endpoint_init()` and `mbedtls_test_move_handshake_to_state()`. For more information have a look at the other tests in `test_suite_ssl.function` and let me know if you have any questions.
2. As mentioned in a comment, I think the best course of action re keeping the `randbytes` buffer would be to only enable the TLS-Exporter when `MBEDTLS_SSL_CONTEXT_SERIALIZATION` is defined. Does that sound okay?

mfil commented on library/ssl_tls.c:
```diff
@@ -9926,4 +9927,117 @@ int mbedtls_ssl_verify_certificate(mbedtls_ssl_context *ssl,
+    }
+
+    prf_input = mbedtls_calloc(prf_input_len, sizeof(unsigned char));
+    label_str = mbedtls_calloc(label_len + 1, sizeof(char));
```
I went for the second alternative.

mfil commented on library/ssl_tls.c:
```diff
@@ -9926,4 +9927,117 @@ int mbedtls_ssl_verify_certificate(mbedtls_ssl_context *ssl,
+    unsigned char *prf_input = NULL;
+    char *label_str = NULL;
+
+    if (use_context && context_len >= (1 << 16)) {
```
Fixed!

mfil commented on library/ssl_tls.c:
```diff
@@ -9926,4 +9927,117 @@ int mbedtls_ssl_verify_certificate(mbedtls_ssl_context *ssl,
+     * If a context is provided, this is then followed by the context length
+     * as a 16-bit big-endian integer, and then the context itself. */
+    memcpy(prf_input, ssl->transform->randbytes + 32, 32);
+    memcpy(prf_input + 32, ssl->transform->randbytes, 32);
```
Fixed!

mfil commented on library/ssl_tls.c:
```diff
@@ -9926,4 +9927,117 @@ int mbedtls_ssl_verify_certificate(mbedtls_ssl_context *ssl,
+        prf_input[65] = (unsigned char) (context_len & 0xff);
+        memcpy(prf_input + 66, context, context_len);
+    }
+    ret = tls_prf_generic(hash_alg, ssl->session->master, 48, label_str,
```
Fixed!

mfil commented on library/ssl_tls.c:
```diff
@@ -9926,4 +9927,117 @@ int mbedtls_ssl_verify_certificate(mbedtls_ssl_context *ssl,
+    const size_t hash_len = PSA_HASH_LENGTH(hash_alg);
+    const unsigned char *secret = ssl->session->app_secrets.exporter_master_secret;
+
+    if (key_len > 0xffff || label_len > 250) {
```
Done!

mfil commented on library/ssl_tls.c:
```diff
@@ -9926,4 +9927,117 @@ int mbedtls_ssl_verify_certificate(mbedtls_ssl_context *ssl,
+                                                    const int use_context)
+{
+    int ret = 0;
+    size_t prf_input_len = use_context ? 64 + 2 + context_len : 64;
```
Done!

mfil commented on library/ssl_tls.c:
```diff
@@ -9926,4 +9927,117 @@ int mbedtls_ssl_verify_certificate(mbedtls_ssl_context *ssl,
+                                                    const int use_context)
+{
+    int ret = 0;
+    size_t prf_input_len = use_context ? 64 + 2 + context_len : 64;
```
Done!

mfil commented on library/ssl_tls.c:
```diff
@@ -9926,4 +9927,117 @@ int mbedtls_ssl_verify_certificate(mbedtls_ssl_context *ssl,
+    memcpy(prf_input + 32, ssl->transform->randbytes, 32);
+    if (use_context) {
+        prf_input[64] = (unsigned char) ((context_len >> 8) & 0xff);
+        prf_input[65] = (unsigned char) (context_len & 0xff);
```
Done!

mfil commented on library/ssl_tls.c:
```diff
@@ -9926,4 +9927,117 @@ int mbedtls_ssl_verify_certificate(mbedtls_ssl_context *ssl,
+    if (use_context) {
+        prf_input[64] = (unsigned char) ((context_len >> 8) & 0xff);
+        prf_input[65] = (unsigned char) (context_len & 0xff);
+        memcpy(prf_input + 66, context, context_len);
```
I defined `const randbytes_len = MBEDTLS_CLIENT_HELLO_RANDOM_LEN + MBEDTLS_SERVER_HELLO_RANDOM_LEN` to make this more readable.

mfil commented on programs/ssl/ssl_server2.c:
```diff
@@ -610,6 +616,8 @@ struct options {
     int event;                  /* loop or event-driven IO? level or edge triggered? */
     uint32_t read_timeout;      /* timeout on mbedtls_ssl_read() in milliseconds    */
+    const char *exp_label;      /* label to input into mbedtls_ssl_export_keying_material() */
+    int exp_len;                /* Lenght of key to export using mbedtls_ssl_export_keying_material() */
```
Fixed!

mfil commented:
@davidhorstmann-arm Thank you for the feedback! I could have used fewer magic numbers.

I think I addressed all the comments on the code. I will add some negative tests later.

davidhorstmann-arm commented:
> I think I addressed all the comments on the code. I will add some negative tests later.

Excellent, thanks! I will take a look tomorrow and hopefully have an answer re the best error code and new config option.

davidhorstmann-arm commented on library/ssl_tls.c:
```diff
@@ -9926,4 +9927,117 @@ int mbedtls_ssl_verify_certificate(mbedtls_ssl_context *ssl,
+                                       const int use_context)
+{
+    if (!mbedtls_ssl_is_handshake_over(ssl)) {
+        return MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
```
I've had a look around and it seems like `MBEDTLS_ERR_SSL_BAD_INPUT_DATA` being used like this has precedent [here](https://github.com/Mbed-TLS/mbedtls/blob/development/include/mbedtls/ssl.h#L5353) and [here](https://github.com/Mbed-TLS/mbedtls/blob/development/include/mbedtls/ssl.h#L5435). In both of these cases the error is used to signal that the caller should wait for the handshake to finish.

With that in mind I'm happy to follow existing convention and leave this as-is. Since it's a bit unusual, could you specifically document this case in the docstring? Something like:
```
\return         0 on success, or
                MBEDTLS_ERR_BAD_INPUT_DATA, if the handshake has not completed yet, or
                Another SSL specific error in case of a different failure.
```

gilles-peskine-arm replied:
@davidhorstmann-arm I share your feeling that `BAD_INPUT_DATA` is not the right error code here. It would be ok for an internal function where it's a coding error to ever invoke this function without knowing the handshake state, but it feels icky in TLS export, especially since it could be called unknowingly during renegotiation.

I don't see another error code that looks really appropriate. Maybe we should create one, and move the other cases you found to that error code in 4.0?

mfil replied:
> I don't see another error code that looks really appropriate. Maybe we should create one, and move the other cases you found to that error code in 4.0?

@gilles-peskine-arm Do you mean I should add the error code in this merge request, and you move the other cases to it later? Or should I still use `BAD_INPUT_DATA`, and you change it later?

davidhorstmann-arm reviewed (changes requested):
LGTM apart from a couple of small nits.

Otherwise remaining things for this to be approved are:
1. Negative testing of the export function in 1.2 and 1.3. If you'd like more guidance I can come up with a list of negative cases to test.
2. It would be good to have a testcase for the example programs if you have time. See `ssl-opt.sh` for what that would look like and let me know if you'd like some help with how to write a testcase.
3. (On me) figure out what to do about the error return value if the handshake isn't finished and get back to you.

mfil commented on ChangeLog.d/add-tls-exporter.txt:
```diff
@@ -0,0 +1,4 @@
+Features
+   * Add the function mbedtls_ssl_export_keying_material() which allows the
+     client and server to extract additional shared symmetric keys from an SSL
+     session, according to the TLS-Exporter specification in RFC 8446 and 5705.
```
Done!

mfil commented on library/ssl_misc.h:
```diff
@@ -1161,10 +1161,11 @@ struct mbedtls_ssl_transform {
+     * See ssl_tls12_populate_transform() */
     unsigned char randbytes[MBEDTLS_SERVER_HELLO_RANDOM_LEN +
                             MBEDTLS_CLIENT_HELLO_RANDOM_LEN];
     /*!< ServerHello.random+ClientHello.random */
```
Done!

davidhorstmann-arm reviewed (commented):
LGTM except for testcases. Would you like a hand with adding more testing?

davidhorstmann-arm commented:
(CI failures are infra-related/spurious)

mfil commented:
@davidhorstmann-arm I added some tests, mostly negative tests but I also added tests to check that the client and server export the same keys.

One of them is currently failing: **TLS 1.3 Keying Material Exporter: Consistent results, large keys**

The problem is coming from lines 165-168 in `mbedtls_ssl_tls13_hkdf_expand_label()` in ssl_tls13_keys.c.

```
    if (buf_len > MBEDTLS_SSL_TLS1_3_KEY_SCHEDULE_MAX_EXPANSION_LEN) {
        /* Should not happen, as above. */
        return MBEDTLS_ERR_SSL_INTERNAL_ERROR;
    }
```

OpenVPN wants to export 256 **bytes** in one go, so this is a problem. I'll check if I can make this function work with larger key sizes.

davidhorstmann-arm commented:
> OpenVPN wants to export 256 bytes in one go, so this is a problem. I'll check if I can make this function work with larger key sizes.

The limitation to 255 bytes seems to date back to #3568. It looks like it's just a case of changing `ssl_tls13_hkdf_encode_label()` to properly encode multi-byte lengths in the HkdfLabel struct, but it's remotely possible that there is some other reason for this limitation. I'll ask one of the reviewers of that PR if they have any idea.

mfil commented:
I fixed the problem with `mbedtls_ssl_tls13_hkdf_expand_label()`. Also, I just realized that the HKDF-Expand algorithm produces at most `255 * hash_size` bytes of output, so I changed the input validation in `mbedtls_ssl_tls13_export_keying_material()` to reflect that.

The tests are working on my machine now, I'll look into test cases for the example programs.

davidhorstmann-arm commented:
I will kick off the CI and take a look at this probably some time next week.

mfil commented:
@davidhorstmann-arm I tried to fix the CI failures, please run the tests again.

mfil commented:
I don't know what these errors mean:

On **Windows-2013-Release-Win32-cmake** (and others): "A volatile slot has not been closed properly" 

On ** all_u16-build_arm_linux_gnueabi_gcc_arm5vte** and others:
```
[2024-10-30T13:33:58.436Z] suites/test_suite_ssl.function: In function 'test_ssl_tls_exporter_rejects_bad_parameters':
[2024-10-30T13:33:58.436Z] suites/test_suite_ssl.function:5277:5: error: 'context' may be used uninitialized in this function [-Werror=maybe-uninitialized]
[2024-10-30T13:33:58.436Z]      mbedtls_free(context);
[2024-10-30T13:33:58.436Z]      ^
[2024-10-30T13:33:58.436Z] suites/test_suite_ssl.function:5276:5: error: 'label' may be used uninitialized in this function [-Werror=maybe-uninitialized]
[2024-10-30T13:33:58.436Z]      mbedtls_free(label);
[2024-10-30T13:33:58.436Z]      ^
[2024-10-30T13:33:58.436Z] suites/test_suite_ssl.function:5275:5: error: 'key_buffer' may be used uninitialized in this function [-Werror=maybe-uninitialized]
[2024-10-30T13:33:58.436Z]      mbedtls_free(key_buffer);
[2024-10-30T13:33:58.436Z]      ^
```
I don't know what it wants from me. The pointers are initialized to NULL. After the allocation, I `memset` the content to 0. (Looking at `TEST_CALLOC`, this should be redundant anyway.) What's uninitialized here?

On **all_u22-test_clang_latest_opt** and others: Too much to paste here, it's having some complaint about `MD_OR_USE_PSA_INIT()`?

Also, how can I reproduce these errors on my machine?

mfil commented:
I think I fixed the second point in my previous comment. I used the macro `MD_OR_USE_PSA_INIT()`, which may jump to the `exit` label, before declaring and initializing the variables. I don't know if that fixes the other errors too.

gilles-peskine-arm commented:
> "A volatile slot has not been closed properly"

A volatile slot is a key created through the PSA API, excluding persistent keys. This error occurs when a test creates a key but does not free it before calling `PSA_DONE()`. It would typically not happen only on Windows. Maybe it would happen elsewhere but it was hidden by the other errors?

The other errors you mention have a common cause: macros like `PSA_INIT`, `TEST_ASSERT` and so on jump to the `exit` label in case of failure. So every variable that's used in the cleanup code needs to be initialized before the first call to one of these macros. Your fix is right, let's see what the CI says.

Jobs using armcc are currently broken on OpenCI (but expected to pass on Internal CI). Please ignore these failures while we work on it.

mfil commented:
> A volatile slot is a key created through the PSA API, excluding persistent keys. This error occurs when a test creates a key but does not free it before calling `PSA_DONE()`.

Could this be because I'm calling `MD_OR_USE_PSA_DONE()` before freeing the endpoints?

It's happening on non-Windows systems too, but not on all. How can I reproduce these errors on my machine?

gilles-peskine-arm commented:
We have a brand new [CI failure FAQ](https://mbed-tls.readthedocs.io/en/latest/kb/testing/understanding-ci-failures/). At a glance, all configurations with both `MBEDTLS_USE_PSA_CRYPTO` and TLS exporter enabled are failing because there's a key that isn't destroyed when the key store is shut down.

> Could this be because I'm calling MD_OR_USE_PSA_DONE() before freeing the endpoints?

Well, yes, that would explain it. `PSA_DONE` (and variants) can't be called while a key is live, and the TLS session uses keys. Typically `PSA_DONE` is only called during cleanup.

mfil commented:
I reproduced the error by enabling `MBEDTLS_USE_PSA_CRYPTO` and fixed it.

That uncovered a new error: In the test `TLS 1.2 Keying Material Exporter: Consistent results, large keys`, I try to export keys of length `UINT16_MAX`. This works when `MBEDTLS_USE_PSA_CRYPTO` is disabled.

When it's enabled, it fails at `psa_key_derivation_tls12_prf_generate_next_block` in psa_crypto.c:
```
    /* We can't be wanting more output after block 0xff, otherwise
     * the capacity check in psa_key_derivation_output_bytes() would have
     * prevented this call. It could happen only if the operation
     * object was corrupted or if this function is called directly
     * inside the library. */
    if (tls12_prf->block_number == 0xff) {
        return PSA_ERROR_CORRUPTION_DETECTED;
    }
```
Apparently, the capacity check did not prevent it. And I don't think the capacity check is wrong because in principle the TLS1.2 PRF can produce arbitrarily large amounts of output. I've made the key size in the large key test somewhat smaller so now all tests should succeed.

I'm thinking that I should just set 255 * 32 = 8160 bytes as a global limit on the exported key size (regardless of TLS version, hash function, etc.). That ought to be enough for anybody.

gilles-peskine-arm commented:
@mfil Thanks for raising that point about TLS12_PRF. Looking into it, that was a bug in our implementation. Our first KDF was HKDF and a few things about it leaked into other KDF. I've filed it as https://github.com/Mbed-TLS/mbedtls/issues/9744. Given the limited impact, I've added it to the backlog.

gilles-peskine-arm commented:
> Is it expected that segfaults can happen when I try to use endpoints without adding these dependencies?

That depends on the nature of the segfault. If the segfault is in the library and it's called with valid parameters (e.g. not a null pointer where a valid context pointer is expected), that's bad. If the segfault is in the test code (e.g. it calls some library function without checking for failures), that may be acceptable, but we'd normally expect a test failure, not a segfault.

I'm not very familiar with the SSL test functions. What I'd normally do to investigate a segfault of unknown origin is either see what ASan+UBSan has to say (`cmake -D CMAKE_BUILD_TYPE=ASan`), or look in a debugger.

mfil commented:
The segfaults happened because I didn't zeroize the enpdoints before using them. After zeroizing, I got regular errors instead of segfaults. But with the `depens_on`s I added in the previous commit, neither should happen.

mfil commented:
@gilles-peskine-arm @davidhorstmann-arm  I added tests for the example programs. Can you kick off the CI again?

davidhorstmann-arm commented:
> Can you kick off the CI again?

Certainly, that should now be running!

mfil commented:
I forgot to call free in the example programs. Should be fixed now.

davidhorstmann-arm commented:
> Do I need a specific version or something?

Correct. The `ssl-opt.sh` testcases are unfortunately very dependent on the specific versions of libraries.

You can get the docker image used in the CI to reproduce it, see [the quickstart guide](https://github.com/Mbed-TLS/mbedtls-test?tab=readme-ov-file#quick-start). Let me know if you would like any help setting it up.

mfil commented:
Thanks @davidhorstmann-arm , I got the docker image working and I found the reasons for the failures:

1) My openssl version is more tolerant than yours, it accepts `-keymatexport=label` while yours wants a space instead of `=`. Fixed!
2) Your openssl version doesn't speak TLS 1.3, so my compatibility test for TLS 1.3 can't work. I removed this test for now.

Is there some way to skip a test when the openssl version is too old? I think it would be nice to keep the test for TLS 1.3 in when running locally. Checking that our implementation agrees with itself isn't that much of a test.

davidhorstmann-arm commented:
>     2. Your openssl version doesn't speak TLS 1.3, so my compatibility test for TLS 1.3 can't work. I removed this test for now.

Ah yes, in fact there is a fix for that too.

Apologies, it's not very well documented but we have multiple versions. Older OpenSSL (`O_CLI`, `O_SRV`) and a newer OpenSSL (`O_NEXT_CLI`, `O_NEXT_SRV`). If you use the `_NEXT` variants and then call the `requires_openssl_tls1_3_with_compatible_ephemeral` helper before the test case that should fix it.

mfil commented:
I added the TLS 1.3 test back in, with `O_NEXT_CLI`. Let's do another CI run!

mfil commented:
I need some help fixing `test_tls13_only_psk`. The TLS 1.3 Exporter compatibility test is failing. To try to fix this, I added `requires_openssl_tls1_3_with_compatible_ephemeral` to the TLS 1.3 exporter compatibility test like you said. (On my machine; in my latest commit, it still has `requires_openssl_next` instead.) Now when I run `ssl-opt.sh`, the test is skipped, but when I run `all.sh test_tls13_only_psk` it still runs and fails.

This is all inside the docker container.

mfil commented:
I think I fixed it by also adding `requires_config_enabled MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED`. Please start another CI run!

davidhorstmann-arm commented:
> I think I fixed it by also adding `requires_config_enabled MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED`. 

I was just trying to chase that up myself. It seems OpenSSL was not happy with it somehow, but I'm not sure why.

> Please start another CI run!

Certainly, will do!

mfil commented:
@davidhorstmann-arm Ok, the last comment that is still open is about returning `MBEDTLS_ERR_SSL_BAD_INPUT_DATA` when the handshake is not yet complete. Have you made a decision about what to do? Should I just document it in the doxygen comment?

davidhorstmann-arm commented:
@mfil Yes, just add it to the doxygen comment please.

The work on adding a new error code would involve some refactoring of existing code, which would stretch the scope of this PR. I can create a followup PR to do this refactoring after this one is merged.

davidhorstmann-arm reviewed (approved):
LGTM, thanks for all your work on this!

Next I believe @waleed-elmelegy-arm will do a second review and then we can merge this.

mfil commented:
@davidhorstmann-arm Sorry that I only saw it after your review, I just found (and fixed) an off-by-1 error in the check for the label length.

The label field in the HkdfLabel struct is up to **255** bytes, including the "tls13 " prefix. So that leaves up to 249 bytes for the label proper.

davidhorstmann-arm reviewed (changes requested):
Looks good to me, but since the problem managed to get through my review, would you be able to add some known-answer testcases for the following?:

* Maximum-length label
* Zero-length label
* Maximum-length context
* Zero-length context

I'm happy if you generate them with the current code, they will just guard against regressions in future and reduce the chance that we reintroduce similar problems.

You should be able to use `mbedtls_printf()` within the testsuite and then run it with `--verbose` from the `tests/` directory to see the output, which will allow you to generate some testcases using a modified version of the test itself.

davidhorstmann-arm commented:
In fact thinking about it an even better way to generate known-answers would be:

1. Modify `ssl_server2.c` to spit out the master secret
2. Perform a handshake with OpenSSL with the `-keymatexport` option set
3. Grab the master secret from `ssl_server2` and the exported key material from OpenSSL as a known-answer testcase. That way we aren't relying on this implementation to be correct.

mfil commented:
I can add more KATs for `ssl_tls13_exporter()` if you mean that function. For `mbedtls_ssl_export_keying_material()`, KATs are difficult because I can't run these without having done a handshake.

I'm not sure how this would help if the label length check is wrong. Let's say I have a KAT with a label of 249 bytes. Someone comes along and raises the limit. The KAT still tests a 249-byte label and succeeds. I *have* added a test that checks if `mbedtls_ssl_export_keying_material()` returns a `BAD_INPUT_ERROR` on a label of length 250.

About `context`, there is no maximum length (well, `SIZE_T_MAX`) in TLS 1.3, because only a hash of it is placed in the HkdfLabel.

mfil commented:
@davidhorstmann-arm Thanks for prodding me to add these tests! I found another bug. It's not that the output is incorrect, but I got an error when using a 249-byte label. Turns out that `mbedtls_ssl_tls13_hkdf_expand_label()` errors if the label is longer than any label that occurs in the key schedule. I changed it so that now the full length may be  used.

davidhorstmann-arm commented:
> Thanks for prodding me to add these tests!

No problem at all, I'm always happy to ask for more tests :)

> Regarding the context length, as I said, there's no maximum because it's hashed before being put in the label. So I don't know what to test there.

Ah I see, sorry I had missed that. In that case, no need to test it then.

davidhorstmann-arm reviewed (changes requested):
I'm afraid I noticed a couple of things that need updating due to the following changes:
* Removing the 255-byte limit on exported material - needs a code comment update
* Allowing externally-supplied labels and output key lengths - we can't use an internal error when they are too big.

mfil commented on library/ssl_tls13_keys.c:
```diff
@@ -150,7 +148,7 @@ int mbedtls_ssl_tls13_hkdf_expand_label(
 
-    if (label_len > MBEDTLS_SSL_TLS1_3_KEY_SCHEDULE_MAX_LABEL_LEN) {
+    if (label_len > MBEDTLS_SSL_TLS1_3_HKDF_LABEL_MAX_LABEL_LEN) {
         /* Should never happen since this is an internal
```
Isn't this still an internal error? The function isn't directly called by the library user and we're checking the length before calling it.

davidhorstmann-arm replied:
Oops sorry I missed the other check on my re-review as I was looking for it in `mbedtls_ssl_tls13_key_exporter()`. Yes this is still an internal error for the label length, so that part is okay as it is.

I believe it is still possible to trigger an internal error by requesting too much keying material though, unless I have missed something there as well?

mfil replied:
No, there's a check for that in `mbedtls_ssl_export_keying_material`.

```
    if (key_len > MBEDTLS_SSL_EXPORT_MAX_KEY_LEN) {
        return MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
    }
```

I have to admit, it's a bit messy because some limits are different for TLS 1.2 and 1.3. But for the exported key length, I've decided to use one global limit that works in all cases because that can even depend on the ciphersuite, and I don't want to deal with that.

davidhorstmann-arm replied:
Ah that's a good point, I had missed that. However, I would like to have a close look to see if this is correct. I'm unsure whether this is a sufficient lower bound, for example in TLS 1.2 where possibly a shorter hash function may be used. I will take a look and get back to you.

mfil replied:
TLS 1.2 uses a different PRF that *technically* can produce unlimited output but #9744 exists, so when using PSA there might indeed be a problem.

davidhorstmann-arm replied:
I think happening across an existing underestimating bug is okay. The problem would be if we tried to generate more than the underlying PRF allows.

davidhorstmann-arm reviewed (approved):
LGTM, thanks for your work and patience!

mfil commented:
@waleed-elmelegy-arm Do you have time to do the second review soon? I see that there's a merge conflict again, but I'd rather wait with fixing it until it's about to get reviewed.

waleed-elmelegy-arm commented:
> @waleed-elmelegy-arm Do you have time to do the second review soon? I see that there's a merge conflict again, but I'd rather wait with fixing it until it's about to get reviewed.

Yes Sorry I will review it shortly.

mfil commented:
I don't understand this test result. Everything's green before results-analysis. In the console output, it says that the tests I added to the suite weren't executed.

waleed-elmelegy-arm commented:
> I don't understand this test result. Everything's green before results-analysis. In the console output, it says that the tests I added to the suite weren't executed.

It appers to be because of 'MBEDTLS_PKCS1' in
`depends_on:MBEDTLS_SSL_PROTO_TLS1_3:MBEDTLS_TEST_AT_LEAST_ONE_TLS1_3_CIPHERSUITE:MBEDTLS_PKCS1`

It is not defined so this test never gets to run.

mfil commented:
@waleed-elmelegy-arm I take it that `MBEDTLS_PKCS1` has been removed. I removed it as a dependency from my tests, I hope this isn't going to break any.

mfil commented:
@davidhorstmann-arm @waleed-elmelegy-arm I tried fixing the test requirements, please do another run.

mfil commented:
I don't understand the CI result. The only error is in the result-analysis step. What's wrong?

gilles-peskine-arm commented on tests/suites/test_suite_ssl.data:
```diff
@@ -3361,3 +3376,67 @@ tls13_srv_max_early_data_size:TEST_EARLY_DATA_HRR:3:3
+ssl_tls_exporter_too_early:MBEDTLS_SSL_VERSION_TLS1_2:1:MBEDTLS_SSL_SERVER_CERTIFICATE
+
+TLS 1.3 Keying Material Exporter: Consistent results, no context
+depends_on:MBEDTLS_SSL_PROTO_TLS1_3:MBEDTLS_TEST_AT_LEAST_ONE_TLS1_3_CIPHERSUITE:MBEDTLS_PKCS1_V15:MBEDTLS_PKCS1_V21:MBEDTLS_ECP_C:MBEDTLS_X509_RSASSA_PSS_SUPPORT
```
[Result analysis](https://mbedtls.trustedfirmware.org/blue/organizations/jenkins/mbed-tls-pr-head/detail/PR-9421-head/80/pipeline/10807) complains

```
[2025-01-27T22:42:54.073Z] *** Analyze driver test_psa_crypto_config_accel_ecc_ecp_light_only vs reference test_psa_crypto_config_reference_ecc_ecp_light_only ***
[2025-01-27T22:42:54.073Z] 
[2025-01-27T22:42:54.073Z] Error: SKIP/FAIL -> PASS: test_suite_ssl;TLS 1.3 Keying Material Exporter: Uses label
…
```

There's part of the explanation in the [CI failure FAQ](https://mbed-tls.readthedocs.io/en/latest/kb/testing/understanding-ci-failures/#outcome-analysis-failed), but it's missing part of the story.

The complaint is that some test cases are skipped in a build with a crypto driver, but passing in a similar build with build-in crypto. We normally expect test coverage parity (using a driver should be transparent), hence the complaint.

A common cause for coverage disparity is if you have legacy dependencies for test cases (or test functions). Here you have dependencies on several legacy crypto configuration options: `MBEDTLS_PKCS1_V15`, `MBEDTLS_PKCS1_V21`, `MBEDTLS_ECP_C`. In the `development` branch, these should all be replaced by the applicable `PSA_WANT_xxx`. In the `mbedtls-3.6` branch, for TLS 1.3, I think PSA dependencies work as well. For TLS 1.2, it's more complicated and ad hoc, because TLS 1.2 in Mbed TLS 3.6 can either use PSA crypto interfaces 
or legacy interfaces depending on `MBEDTLS_USE_PSA_CRYPTO`.

mfil replied:
I've tried copying the dependencies from "Handshake, tls1_3" since all that should be necessary for these tests is that you can successfully do a TLS 1.3 handshake. That still contains `MBEDTLS_PKCS1_V21` though...

Also fixed the latest merge conflict

mfil commented:
The failing test is all_u16-test_depends_py_pkalgs

```
[2025-02-11T22:55:24.038Z] depends.py: !MBEDTLS_X509_RSASSA_PSS_SUPPORT PASSED
[2025-02-11T22:55:24.038Z] depends.py: !MBEDTLS_ECDSA_C !MBEDTLS_ECP_C !MBEDTLS_PKCS1_V21 !MBEDTLS_RSA_C !MBEDTLS_X509_RSASSA_PSS_SUPPORT passed
[2025-02-11T22:55:24.038Z] depends.py: !MBEDTLS_PKCS1_V15 FAILED
[2025-02-11T22:55:24.038Z] ^^^^test_depends_py_pkalgs: test/build: depends.py pkalgs (gcc): tests/scripts/depends.py pkalgs -> 1^^^^
[2025-02-11T22:55:24.038Z] make clean
```

But `./tests/scripts/mbedtls-all.sh test_depends_py_pkalgs` works on my machine.

gilles-peskine-arm commented:
@mfil The CI uses an old version of GCC, but it shouldn't make a difference, and I can reproduce the failure on my machine (Ubuntu 24.04). A few tips for easier reproduction:

* Run `tests/scripts/depends.py -e '!MBEDTLS_PKCS1_V15'` to reproduce the failing build and keep the configuration and build products.
* Change `TEST_ASSERT(ret == 0)` to `TEST_EQUAL(ret, 0)` to see the value of `ret` on failure.
* If you want to build with different options (e.g. debug or Asan), edit the `CFLAGS` setting in `tests/scripts/depends.py`. Or run `depends.py -e` and let it fail, then `make clean` and rebuild with different flags.

mfil commented:
@gilles-peskine-arm The tests didn't fail because I was on the wrong branch.

I think I found the appropriate `PSA_WANT_*` dependencies to make it work.

mfil commented:
Also, since most of my commits are fiddling with test dependencies, should I just squash them all together when it works?

gilles-peskine-arm commented:
@mfil We generally prefer to keep the development history. But it's ok to squash a series of commits that are “boring”, such as fiddling with the same thing until the CI passing, or fixing simple typos. We would prefer to keep the history in cases like {commit 1: general case; commit 2: fix subtle bug in one configuration}. Thanks for enquiring!

gilles-peskine-arm commented:
Speaking of commits... I see 60 commits, 14 files changed, `size-m`. Going by experience, that's a sign that review will be very difficult. 60 commits, even if many are “boring” commits, is outside our comfort range and tends to result in either merging broken code due to review fatigue, or endless succession of rework.

Is there a sensible way to split this pull request into 2 or 3 smaller pieces? For example preliminary refactorings, then TLS 1.2, then TLS 1.3. Or create internal functions with unit tests, then finish integration and add support in `ssl*2.c`. Or whatever makes sense, I'm basing these suggestions only on a very quick glance at which files are getting changed.

@davidhorstmann-arm You reviewed an earlier version, what's your opinion on the complexity of this PR? Would you be ok if it gets split?

davidhorstmann-arm commented:
> @davidhorstmann-arm You reviewed an earlier version, what's your opinion on the complexity of this PR? Would you be ok if it gets split?

I think it's fine to stay as it is, squashing some of the 'boring' commits would also be fine. Most of the commits are fiddling with the tests and I can run through them quite easily when reviewing.

The feature itself is very small and quite easy to review. Most of the extra files-changed are either testing or configuration-related.

davidhorstmann-arm commented:
I would prefer to keep it as-is rather than splitting for 2 reasons:
1. This PR is fairly far along and I am quite close to approving it.
2. Most of the extra size is testing and I'd rather merge that in one with the feature itself.

davidhorstmann-arm commented on tests/suites/test_suite_ssl.data:
```diff
@@ -3333,3 +3348,67 @@ tls13_srv_max_early_data_size:TEST_EARLY_DATA_HRR:3:3
+ssl_tls_exporter_too_early:MBEDTLS_SSL_VERSION_TLS1_2:1:MBEDTLS_SSL_SERVER_CERTIFICATE
+
+TLS 1.3 Keying Material Exporter: Consistent results, no context
+depends_on:MBEDTLS_SSL_PROTO_TLS1_3:MBEDTLS_TEST_AT_LEAST_ONE_TLS1_3_CIPHERSUITE:MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED:PSA_WANT_ALG_RSA_PKCS1V15_SIGN:PSA_WANT_ALG_RSA_PKCS1V21_SIGN
```
Ah I see what's going on - result analysis is failing because there is no `PSA_WANT_ALG_RSA_PKCS1V21_SIGN` so these cases are never executed.

This can be replaced with a dependency on `PSA_WANT_ALG_RSA_PSS` instead.

mfil replied:
It seems that `depends_on:MBEDTLS_SSL_PROTO_TLS1_3:MBEDTLS_TEST_AT_LEAST_ONE_TLS1_3_CIPHERSUITE:MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED:PSA_WANT_ALG_RSA_PKCS1V15_SIGN:MBEDTLS_X509_RSASSA_PSS_SUPPORT` makes ./tests/scripts/mbedtls-all.sh test_depends_py_pkalgs pass and doesn't prevent the tests from running at all.

That's what I have in my latest commit. Does this make sense?

Replacing the PKCS one with `PSA_WANT_ALG_RSA_PSS` causes tests to fail.

davidhorstmann-arm replied:
Hmm okay. I'll kick off the CI while I think about that more carefully..

gilles-peskine-arm replied:
There are already TLS 1.3 test cases with dependencies on both `MBEDTLS_PKCS1_V21` and `MBEDTLS_X509_RSASSA_PSS`. I didn't expect that: I thought TLS 1.3 would rely on PSA for signatures, which means it should use PSA dependencies.

Normally, in Mbed TLS ≤3.6, the test dependencies must match the API that is used to perform a cryptographic operations, because it's possible to build without the PSA subsystem (so the cryptographic mechanism is only available though legacy APIs) and it's also possible to build with PSA drivers (so the mechanism is only available through PSA APIs). But for RSA, we don't support driver-only builds, so using legacy dependency should still work. In the development branch, however, PSA APIs should always work except in low-level cryptography modules.

So for now, I'd expect that either `MBEDTLS_PKCS1_V21` or `PSA_WANT_ALG_RSA_PSS` would work, and `PSA_WANT_ALG_RSA_PSS` would be preferred. I don't understand why `MBEDTLS_PKCS1_V21` isn't working here when it's working in other tests.

mfil commented:
@davidhorstmann-arm @waleed-elmelegy-arm Tests are all working now. Should I go ahead and squash the commits that tweak the test dependencies?

davidhorstmann-arm commented:
@mfil That sounds good to me, I'll wait for that and then start re-reviewing.

mfil commented:
I see no failures in the interface stability tests.

gilles-peskine-arm commented:
The root cause for the failed status of the [interface stability tests job](https://mbedtls.trustedfirmware.org/job/mbed-tls-pr-merge/job/PR-9421-merge/150/console) is:

```
Superseded by mbed-tls-pr-merge/PR-9421-merge#151
```

(It's not easy to find in the logs!)

The [referenced job](https://mbedtls.trustedfirmware.org/job/mbed-tls-pr-merge/job/PR-9421-merge/151/console) failed because
```
PR author not found on allowlist - not building
```

That is expected: we don't automatically run tests on external contributions, a human needs to do a cursory review first for security reasons.

What is not expected is that there was a race condition. Normally we'd have expected the automatic 150 to have finished by the time a human has started 151.

I'm not sure that happens enough to matter.

For this one time, I've started a new job.

davidhorstmann-arm reviewed (approved):
LGTM, thanks! (Apologies for the delay.)

I'm afraid there is a small conflict in `test_suite_ssl.function` because another PR has added tests there as well. Would you be able to resolve this conflict? It shouldn't be too difficult. Once that's done I'll try to get this re-approved quickly so we can merge.

mfil commented:
@davidhorstmann-arm @waleed-elmelegy-arm The merge conflict is resolved!

davidhorstmann-arm commented on tests/suites/test_suite_ssl.function:
```diff
@@ -5198,5 +5229,263 @@ exit:
     mbedtls_debug_set_threshold(0);
     mbedtls_free(first_frag);
     PSA_DONE();
+/* END_CASE */
```
The way this conflict got resolved is wrong I think - just needs a closing `}` by the looks of it.

I think this is because of the way git handles the conflict poorly.

davidhorstmann-arm reviewed (changes requested):
One problem with the conflict resolution, would you be able to redo the rebase? LGTM otherwise.

davidhorstmann-arm reviewed (approved):
LGTM, thanks! I'll start the CI

mfil commented:
@waleed-elmelegy-arm If you don't have time for a review, would it be possible to assign someone else?

waleed-elmelegy-arm commented:
> @waleed-elmelegy-arm If you don't have time for a review, would it be possible to assign someone else?

Apologies I was off. I am currently reviewing it.

waleed-elmelegy-arm reviewed (approved):
LGTM, Thank you very much.

gilles-peskine-arm commented:
There's a 3.6 backport but it isn't up to date. @mfil  @davidhorstmann-arm @waleed-elmelegy-arm Is there still interest in the 3.6 backport? Given the size of the code now, is it still ok for a long-term support branch?

gilles-peskine-arm commented:
Having checked internally, it seems that we would still like to include this in Mbed TLS 3.6. @mfil Can you please update https://github.com/Mbed-TLS/mbedtls/pull/9469 ? Since it hasn't been reviewed yet, please rebase it on top of the current head of `mbedtls-3.6`, and update the patch to the version that's approved for the `development` branch. If any differences are needed in 3.6, please briefly explain what's changed, so we know what to expect in `git range-diff`.

mfil commented:
The backport pull request is ready!

gilles-peskine-arm reviewed (approved):
Approving the change from 1a1ec2fccee002bb886a960fc0909f29fca3a7dd to dba07e152e60112570773921db89e6fcc6d549f1

davidhorstmann-arm commented:
@mfil Thanks for all your work on this. Everything looks fine and the backport is approved, so I'll add it to the merge queue!
